sklearn/linear_model/_quantile.py

```diff
@@ -218,37 +224,47 @@ def fit(self, X, y, sample_weight=None):
         #
         # Filtering out zero samples weights from the beginning makes life
         # easier for the linprog solver.
-        mask = sample_weight != 0
-        n_mask = int(np.sum(mask))  # use n_mask instead of n_samples
+        indices = np.flatnonzero(sample_weight != 0)
+        n_indices = len(indices)  # use n_mask instead of n_samples
         c = np.concatenate(
             [
                 np.full(2 * n_params, fill_value=alpha),
-                sample_weight[mask] * self.quantile,
-                sample_weight[mask] * (1 - self.quantile),
+                _safe_indexing(sample_weight, indices) * self.quantile,
+                _safe_indexing(sample_weight, indices) * (1 - self.quantile),
             ]
         )
         if self.fit_intercept:
             # do not penalize the intercept
             c[0] = 0
             c[n_params] = 0
 
-            A_eq = np.concatenate(
-                [
-                    np.ones((n_mask, 1)),
-                    X[mask],
-                    -np.ones((n_mask, 1)),
-                    -X[mask],
-                    np.eye(n_mask),
-                    -np.eye(n_mask),
-                ],
-                axis=1,
-            )
-        else:
-            A_eq = np.concatenate(
-                [X[mask], -X[mask], np.eye(n_mask), -np.eye(n_mask)], axis=1
-            )
+        if (sparse.issparse(X)) and (X.format == "coo"):
+            X = X.tocsr()
+        X = _safe_indexing(X, indices)
 
-        b_eq = y[mask]
+        if sparse.issparse(X):
+            sparse_constructor = {
+                "csr": sparse.csr_matrix,
+                "csc": sparse.csc_matrix,
+                "coo": sparse.coo_matrix,
```

Comment:
Let's remove the `coo` as well here.

Reply:
Even better: I think we can stick to CSC only. The highs and interior-point solvers always converts to CSC, see
- highs: https://github.com/scipy/scipy/blob/c9cdbf2804d092eb63dc66e885ca99af11f381e5/scipy/optimize/_linprog_highs.py#L345-L349.
- interior point: https://github.com/scipy/scipy/blob/c9cdbf2804d092eb63dc66e885ca99af11f381e5/scipy/optimize/_linprog_ip.py#L722-L724